### PR TITLE
Ring of Loki Block Selection Work

### PR DESCRIPTION
**Pull request by LewisSaber**
- Added new keybind, when pressed removes selection from all blocks
- Made it so if ring of loki is off selection is not drawn

### CODE REVIEW

File: src/main/java/vazkii/botania/common/item/relic/ItemLokiRing.java [hunk 1]
```diff
@@ -215,13 +215,17 @@ public static void renderHUDNotification(HUD_MESSAGE type){
 			case BREAKING:
 				text = getLokiBreakingModeText(getLokiRing(mc.thePlayer));
 				break;
+			case CLEAR:
+				text = getLokiCearText(getLokiRing(mc.thePlayer));
+			break;
```
Comment (combusterf) [inline src/main/java/vazkii/botania/common/item/relic/ItemLokiRing.java:220]: Indentation

File: src/main/java/vazkii/botania/common/item/relic/ItemLokiRing.java [hunk 2]
```diff
@@ -235,6 +239,11 @@ public static String getLokiBreakingModeText(ItemStack stack){
 				EnumChatFormatting.RED + StatCollector.translateToLocal("botaniamisc.lokiOff"));
 	}
 
+	public static String getLokiCearText(ItemStack stack){
+		return EnumChatFormatting.GOLD + StatCollector.translateToLocal("botaniamisc.lokiClear") ;
+				
+	}
```
Comment (combusterf) [inline src/main/java/vazkii/botania/common/item/relic/ItemLokiRing.java:245]: Various bits of excess whitespace, before the semicolon and at the end of the function.

**Review by combusterf — DISMISSED** [review]
Just a few cleanups to do, looking fine otherwise.

**Comment by combusterf** [comment]
Code style aside, at this rate we seem to bind the entire keyboard to the loki ring. From an user interaction perspective this is rather poor design - it requires you to deconflict keybinds, learn from the settings menu what the features are, and that you move your hand varying distances across the keyboard. 

Wouldn't it be better to limit keybinds to at most one key, and combine it with shift/ctrl/mouse instead?

**Comment by LewisSaber** [comment]
Thats a fair point, I have run out of ideas I can implement, if I think of any and it requires new hotkey, I will try to combine some of them(like turning on/off both mode and breaking)

**Comment by Alastors** [comment]
> Code style aside, at this rate we seem to bind the entire keyboard to the loki ring. From an user interaction perspective this is rather poor design - it requires you to deconflict keybinds, learn from the settings menu what the features are, and that you move your hand varying distances across the keyboard. 
> 
> 
> 
> Wouldn't it be better to limit keybinds to at most one key, and combine it with shift/ctrl/mouse instead?

That, or at this point it might necessitate keybind wheels, like how TC handles foci

**Review by combusterf — COMMENTED** [review]
Code is a-ok now. Feel free to merge once we have consensus on the keybind topic.

**Comment by combusterf** [comment]
My gut feeling says the focus wheel would be a bit overengineered, and shift/ctrl combinations would be fine enough as far as I'm concerned. There's an inherent risk in merging the feature as-is now to never improve it later, so you can address it and I'll get it merged, or if a second opinion comes to approve it.

**Comment by Alastors** [comment]
> My gut feeling says the focus wheel would be a bit overengineered, and shift/ctrl combinations would be fine enough as far as I'm concerned. There's an inherent risk in merging the feature as-is now to never improve it later, so you can address it and I'll get it merged, or if a second opinion comes to approve it.
> 
> 
> 
> (and sorry for the spam, it took some trial and error to remove myself as a blocker)

That's probably fair, would just be future proofing the ring incase people want to add more features, but yeah thats a fair point

**Comment by LewisSaber** [comment]
image

File: src/main/java/vazkii/botania/common/network/PacketLokiToggle.java [hunk 3]
```diff
@@ -25,12 +25,24 @@ public IMessage onMessage(PacketLokiToggle message, MessageContext ctx) {
...
+                ItemLokiRing.setMode(aRing, ringState);
+                PacketLokiToggleAck ackMessage = new PacketLokiToggleAck();
+                ackMessage.state = ringState;
+                PacketHandler.INSTANCE.sendTo(ackMessage, player);
+            }
         }
         return null;
     }
+
+
```
Comment (combusterf) [inline src/main/java/vazkii/botania/common/network/PacketLokiToggle.java:47]: Excess whitespace can be cleaned up

File: src/main/resources/assets/botania/lang/en_US.lang [hunk 4]
```diff
@@ -155,13 +155,20 @@ botaniamisc.altarRefill1=add back the last recipe
...
 botaniamisc.lokiOn=On
 botaniamisc.lokiOff=Off
 botaniamisc.breaking=Breaking
 botaniamisc.lokiClear=Successfully cleared all selection
+botaniamisc.lokiDescription=Builder's wand but §c§oC§6§oo§e§oo§a§ol§9§oe§1§or
+botaniamisc.lokiDescription2=Can Place/Interact/Break with multiple blocks at once!
+botaniamisc.lokiToggleDescription=Mode Switch Key to turn ring
+botaniamisc.lokiBreakingDescription=Mode Switch Key + Shift to toggle breaking
+botaniamisc.lokiCurrent=Current
+botaniamisc.lokiState=Ring
```
Comment (combusterf) [inline src/main/resources/assets/botania/lang/en_US.lang:171]: I didn't see Russian translations for these yet, did you forget them or was that on purpose?

**Review by combusterf — CHANGES_REQUESTED** [review]
I like this way better, just a few cleanups from the new changes and I'll have it merged

File: src/main/resources/assets/botania/lang/en_US.lang (hunk 4 above)
Comment (LewisSaber) [reply src/main/resources/assets/botania/lang/en_US.lang:171]: on purpose, I might do it later, I plan to give ring of loki tooltip more love soon so I will make translation when tooltip will be closer to final

**Comment by LewisSaber** [comment]
This should be it

**Review by combusterf — APPROVED** [review]

**Pull request merged**